resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshotsStoreImpl.scala

```diff
@@ -0,0 +1,95 @@
+ * limitations under the License.
+ */
+package org.apache.spark.scheduler.cluster.k8s
+
+import java.util.concurrent.{ExecutorService, ScheduledExecutorService, TimeUnit}
+
+import com.google.common.collect.Lists
+import io.fabric8.kubernetes.api.model.Pod
+import io.reactivex.disposables.Disposable
+import io.reactivex.functions.Consumer
+import io.reactivex.schedulers.Schedulers
+import io.reactivex.subjects.PublishSubject
+import javax.annotation.concurrent.GuardedBy
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.spark.util.{ThreadUtils, Utils}
+
+private[spark] class ExecutorPodsSnapshotsStoreImpl(
+    bufferSnapshotsExecutor: ScheduledExecutorService,
+    executeSubscriptionsExecutor: ExecutorService)
+  extends ExecutorPodsSnapshotsStore {
+
+  private val SNAPSHOT_LOCK = new Object()
+
+  private val snapshotsObservable = PublishSubject.create[ExecutorPodsSnapshot]()
+  private val observedDisposables = mutable.Buffer.empty[Disposable]
+
+  @GuardedBy("SNAPSHOT_LOCK")
+  private var currentSnapshot = ExecutorPodsSnapshot()
+
+  override def addSubscriber(
+      processBatchIntervalMillis: Long)
+      (subscriber: ExecutorPodsSnapshot => Unit): Unit = {
+    observedDisposables += snapshotsObservable
+      // Group events in the time window given by the caller. These buffers are then sent
+      // to the caller's lambda at the given interval, with the pod updates that occurred
+      // in that given interval.
+      .buffer(
+        processBatchIntervalMillis,
+        TimeUnit.MILLISECONDS,
+        // For testing - specifically use the given scheduled executor service to trigger
+        // buffer boundaries. Allows us to inject a deterministic scheduler here.
+        Schedulers.from(bufferSnapshotsExecutor))
+      // Trigger an event cycle immediately. Not strictly required to be fully correct, but
+      // in particular the pod allocator should try to request executors immediately instead
+      // of waiting for one pod allocation delay.
+      .startWith(Lists.newArrayList(ExecutorPodsSnapshot()))
+      // Force all triggered events - both the initial event above and the buffered ones in
+      // the following time windows - to execute asynchronously to this call's thread.
+      .observeOn(Schedulers.from(executeSubscriptionsExecutor))
+      .subscribe(toReactivexConsumer { snapshots: java.util.List[ExecutorPodsSnapshot] =>
+        Utils.tryLogNonFatalError {
+          snapshots.asScala.foreach(subscriber)
+        }
+      })
+  }
+
+  override def stop(): Unit = {
+    observedDisposables.foreach(_.dispose())
+    snapshotsObservable.onComplete()
+    ThreadUtils.shutdown(bufferSnapshotsExecutor)
+    ThreadUtils.shutdown(executeSubscriptionsExecutor)
+  }
+
+  override def updatePod(updatedPod: Pod): Unit = SNAPSHOT_LOCK.synchronized {
+    currentSnapshot = currentSnapshot.withUpdate(updatedPod)
```

Comment:
So the watch only creates new snapshots by applying some "diff" according to the next event to the previous snapshot. One downside of this data model is that we end up buffering multiple collections of pods in the observable stream which can all perhaps only differ by a single pod per update. Thus we end up temporarily storing redundant information in the snapshots. But the observable buffers are ephemeral and will be periodically processed by the periodic iterations of the subscribers. I wouldn't mind thinking about a more optimal representation here.

Reply:
we only ever need to send the latest snapshot update with this new approach, so having the `snapshotsObservable buffer` doesn't make sense anymore? I.e., we don't need to send the subscriber all the Snapshots that happened during a given interval. Only the last one.
So we can just have a periodic task that takes the latest snapshot and sends it to all subscribers.

Reply:
I'm pretty sure we want all snapshots. The reason is to get a more accurate response in the lifecycle handler. If a pod enters and error state and then someone marks it for deletion, you want to at least have the chance to capture the error state when sending the executor removal request to the Spark scheduler.

For the pods allocator we can do either the latest or all.

Reply:
gotcha, that makes sense

Reply:
A brief remark that one still might not be 100% accurate because of failures in the information flow (missing events etc) - but for all the events we do get we should be trying to handle them all as best we can.